src/main/java/com/minecolonies/coremod/client/gui/WindowBuildingBrowser.java

```diff
@@ -0,0 +1,345 @@
+            {
+                final BuildingInfo building = buildings.get(index);
+                final Text packLabel = rowPane.findPaneOfTypeByID("packName", Text.class);
+                final Text nameLabel = rowPane.findPaneOfTypeByID("buildingName", Text.class);
+                final Text sizeLabel = rowPane.findPaneOfTypeByID("buildingSize", Text.class);
+                final Text levelLabel = rowPane.findPaneOfTypeByID("buildingLevel", Text.class);
+
+                packLabel.setText(Component.literal(building.pack().getName()));
+                nameLabel.setText(Component.literal(building.path()));
+                nameLabel.setColors(building.isParent() ? building.isInvisible() ? COLOR_INVISIBLE_CHILD : COLOR_CHILD
+                        : building.isInvisible() ? COLOR_INVISIBLE : COLOR_NORMAL);
+                sizeLabel.setText(Component.literal(String.format("%d x %d x %d",
+                        building.size().getX(), building.size().getY(), building.size().getZ())));
+                if (building.levels().isEmpty())
+                {
+                    levelLabel.hide();
+                }
+                else
+                {
+                    levelLabel.show();
+                    levelLabel.setText(formatLevels(building.levels()));
+                }
+            }
+        });
+    }
+
+    @NotNull
+    private Component formatLevels(@NotNull final Set<Integer> levels)
+    {
+        final List<Integer> list = new ArrayList<>(levels);
+        if (list.size() == 1)
+        {
+            return Component.literal(Integer.toString(list.get(0)));
+        }
+        final int minLevel = list.get(0);
+        final int maxLevel = list.get(list.size() - 1);
+        if (list.size() == maxLevel - minLevel + 1)
+        {
+            return Component.translatable("%s-%s", Integer.toString(minLevel), Integer.toString(maxLevel));
+        }
+        return Component.literal(String.join(",", list.stream().map(i -> Integer.toString(i)).toList()));
+    }
+
+    @NotNull
+    private List<BuildingInfo> discoverBuildings()
+    {
+        final List<BuildingInfo> buildings = new ArrayList<>();
+        if (!StructurePacks.waitUntilFinishedLoading())
+        {
+            return buildings;
+        }
+
+        // to reduce total search time, kick work off to several worker threads (currently 4 threads ~= 5s on a decent CPU and slow disk)
+        final ExecutorService packPool = Executors.newFixedThreadPool(WORKER_THREADS, runnable ->
+        {
+            final Thread thread = new Thread(runnable, "Minecolonies Building Browser Worker");
+            thread.setDaemon(true);
+            thread.setUncaughtExceptionHandler((thread1, throwable) -> Log.getLogger().error("Minecolonies Building Browser errored! ", throwable));
+            return thread;
+        });
+        final Map<StructurePackMeta, Future<List<BuildingInfo>>> packFutures = StructurePacks.getPackMetas().stream()

```

Comment:
Couldnt we use the IO threadpool minecraft already has here?

Reply:
Possibly.  I didn't find one when I went looking, but perhaps I just overlooked it.  Do you recall where it is?

Reply:
Still not seeing anything on client side.  There is one on server side, but I'm deliberately not going there.

Reply:
public final class IOPool in Structurize should also do the trick tbh

Reply:
The top level does run on structurize's IOPool.  But that's only a single thread, so it couldn't fork for the subtasks, it just got wedged.

Reply:
Minecraft class itself used to be an executor (like at least two versions ago it was), also there was something in util(s) class

Reply:
That's not a threadpool, that makes it do work on the client/render thread, which is the opposite of the goal.